Fix nonequality join predicate pushdown

losipiuk opened the pull request.

losipiuk commented:
@kokosing, @pnowojski please do the first pass

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -458,15 +460,24 @@ private OuterJoinPushDownResult processLimitedOuterJoin(Expression inheritedPred
                 }
             }
 
-            // See if we can push down any outer or join predicates to the inner side
```
@martint please take a look. Here. I am not 100% sure I did not mess something up here.
My intention was to rewrite joinPredicate in context of inner symbols.
Did I break some optimization by removing `joinPredicate` from `EqualityInference.nonInferrableConjuncts(and(outerEffectivePredicate, joinPredicate)` ?

losipiuk replied:
Actually tests still pass if `joinPredicate` in that changed line.
Review from someone with more understanding of what is happening here would be awesome :)

martint commented:
@erichwang, can you take a look, too?

martint commented on presto-main/src/main/java/com/facebook/presto/operator/SimplePagesHashStrategy.java:
```diff
@@ -59,18 +59,19 @@ public SimplePagesHashStrategy(List<Type> types, List<List<Block>> channels, Lis
 
     private List<Block[]> buildChannelArrays(List<List<Block>> channels)
     {
+        List<Block[]> channelArrays = new ArrayList<>();
```
Is there a query that exposes this issue? Can you add a test?

losipiuk replied:
https://github.com/prestodb/presto/pull/5403/commits/c46ccfc8e8a8c8075b8a4c283b5dfc0877d9185b

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -321,7 +321,7 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
                     leftPredicate = leftOuterJoinPushDownResult.getOuterJoinPredicate();
```
Can you explain the purpose of this commit? Maybe add the explanation to the commit message. I can see what it's doing, but I can't tell what it's for or why it's needed.

losipiuk replied:
I think it should just be amended to `Push OUTER JIN condition to INNER side in PredicatePushDown`.
I will let you finish the review before doing that.

losipiuk replied:
done

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -422,6 +422,7 @@ private OuterJoinPushDownResult processLimitedOuterJoin(Expression inheritedPred
 
```
Typo in commit message: JIN -> JOIN

losipiuk replied:
done

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -458,15 +460,24 @@ private OuterJoinPushDownResult processLimitedOuterJoin(Expression inheritedPred
-            // See if we can push down any outer or join predicates to the inner side
-            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(and(outerEffectivePredicate, joinPredicate))) {
+            // See if we can push down any outer predicates to the inner side
+            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(and(outerEffectivePredicate))) {
```
`and` shouldn't be needed since there's only one term now.

losipiuk replied:
Yeah - assuming we actually want to remove joinPredicate from this line. As I wrote I am not super sure what was original intention here.

erichwang replied:
yea we can get rid of the AND there

losipiuk replied:
done

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -194,6 +194,7 @@
 import static com.facebook.presto.sql.planner.SystemPartitioningHandle.SINGLE_DISTRIBUTION;
```
Is there a query that exposes this issue? Add a test?

losipiuk replied:
Actually I noticed issue here in the middle of the work. Later on I added this hack of always adding `TRUE_LITERAL` expression to `JoinNode.filter` for OUTER joins if we end up with empty join condition.
This is temporary hack to force using non-compiled JoinProbe implementation.
This is needed until compiled version can support filter functions.

Side effect of this hack is that expression
```
clauses.isEmpty() && !node.getFilter().isPresent() && node.getType() != INNER
```  
is always false for now.

But there are queries in tests which would fail here if the hack was not present.
E.g. the query I added in fixup commit
```sql
SELECT a FROM (VALUES 1, 2) t1(a) RIGHT OUTER JOIN (VALUES 10, 11) t2(b) ON a > 1
```

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java:
```diff
@@ -58,7 +58,7 @@ public PlanNode visitApply(ApplyNode node, RewriteContext<PlanNode> context)
                 return new JoinNode(
                         idAllocator.getNextId(),
-                        JoinNode.Type.FULL,
+                        JoinNode.Type.INNER,
```
Why is this needed?

losipiuk replied:
This was a bug actually. And worked incidentally because we converted any 
Join without conditions into NestedLoopJoin (which was not correct for OUTER for case where INNER table was empty).

After changing that in LocalExecutionPlanner:

```
+            if (clauses.isEmpty() && !node.getFilter().isPresent() && node.getType() == INNER) {
                  return createNestedLoopJoin(node, context);
              }
```

this stopped working.

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java:
```diff
@@ -243,24 +238,6 @@ protected RelationPlan visitJoin(Join node, Void context)
             for (Expression conjunct : ExpressionUtils.extractConjuncts(criteria)) {
                 conjunct = ExpressionUtils.normalize(conjunct);
 
-                // We skip this section for joins using "USING" clause as generated criteria uses non qualified column names.
```
I think the correct approach here is to drop the entire expression into the JoinNode filterCondition and let the predicate push down extract out the things that are needed. I think this will work with your changes below and one additional change.

losipiuk replied:
I agree 100%. I just wanted this PR to be bugfix rather than refactor. I will do the refactor as last commit in this PR so we can decide if want this in before next release or after.

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -627,6 +636,7 @@ private static Expression extractJoinPredicate(JoinNode joinNode)
             for (JoinNode.EquiJoinClause equiJoinClause : joinNode.getCriteria()) {
                 builder.add(equalsExpression(equiJoinClause.getLeft(), equiJoinClause.getRight()));
             }
+            joinNode.getFilter().ifPresent(filter -> builder.add(filter));
```
filter -> builder.add(filter)
should be 
builder::add

losipiuk replied:
done

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -357,7 +357,9 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
                     rightSource != node.getRight() ||
                     !expressionEquivalence.areExpressionsEquivalent(session, newJoinPredicate, joinPredicate, types)) {
-                if (newJoinPredicate.equals(BooleanLiteral.TRUE_LITERAL)) {
+                if (node.getType() == JoinNode.Type.INNER && newJoinPredicate.equals(BooleanLiteral.TRUE_LITERAL)) {
```
We should drive the code through here if the join predicate is TRUE, but we should retain the node.getType() for the execution so it knows what to do with it

losipiuk replied:
removed

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -374,28 +374,37 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
                             .collect(Collectors.toMap(key -> key, Symbol::toQualifiedNameReference)));
```
We also need to change line 361 here so that node.getFilter is not propagated to the JoinNode now that the join predicate includes the filter from your changes.

losipiuk replied:
no longer relevant

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -1391,7 +1392,7 @@ public PhysicalOperation visitIndexJoin(IndexJoinNode node, LocalExecutionPlanCo
         {
             List<JoinNode.EquiJoinClause> clauses = node.getCriteria();
-            if (clauses.isEmpty() && !node.getFilter().isPresent()) {
+            if (clauses.isEmpty() && !node.getFilter().isPresent() && node.getType() == INNER) {
```
I think the best thing to do here is to modify the nested loop join code to be aware of the type of join so that it can emit nulls correctly.

Otherwise, this will force LJ, RJ, and FJ to use the hash build join even when the criteria is just TRUE. The nested loop join will be much more efficient in those cases.

losipiuk replied:
Yeah. But I would rather have it in separate PR.
Handling OUTER JOIN when build side is OUTER side is not trivial. I can mimic how it was done for lookup joins but this is somewhat complicated.

I do not want this to be part of this PR, which is supposed to be bugfix PR.

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -458,15 +460,24 @@ private OuterJoinPushDownResult processLimitedOuterJoin(Expression inheritedPred
 
-            // TODO: consider adding join predicate optimizations to outer joins
+            // See if we can push down any join predicates to the inner side
+            for (Expression conjunct : extractConjuncts(joinPredicate)) {
```
nice, this block looks correct to me

erichwang commented:
nice! generally the right direction, but a few modifications suggested

losipiuk commented:
@martint, @erichwang PR updated.

The build is still running but hopefully it will finish green.
The case with empty join condition now works. 
I also changed `SimplePagesHashStrategy.buildChannelArrays` and moved logic for handling empty channels list to `applyFilterFunction`.

I did not change the TODO in `PagesIndex` as I somewhat lost what I was actually supposed to change there :/

The difference since Friday PR is here:
https://gist.github.com/losipiuk/2e7814e13b1361a825c850d16aa0096d

erichwang commented on presto-main/src/main/java/com/facebook/presto/operator/SimplePagesHashStrategy.java:
```diff
@@ -200,6 +200,16 @@ public boolean positionEqualsPosition(int leftBlockIndex, int leftPosition, int
+    private Block[] getLeftBlocks(int leftBlockIndex)
+    {
+        if (channelArrays.isEmpty()) {
+            return EMPTY_BLOCK_ARRAY;
```
instead of putting this here, have the applyFilterFunction directly return the boolean if channelArrays.isEmpty()

losipiuk replied:
You mean to precalculate rasult of `filterFunction`. Makes sense :)

losipiuk replied:
Well I take that back.
filterFunction can only use channels from probe side. And I actually need to calculate it.
Am I missing something?

erichwang replied:
i might be missing something too. I'm trying to reconcile how we can construct a filter function that requires something from one of the channels, but provides no channels?

losipiuk replied:
I am thinking of query like this one.

```sql
SELECT t2.a FROM t1 JOIN t2 ON t2.c > t2.d
```


This is dumb query. But technically (if join condition is not pushed down to filter which is not always possible) it does not use any channels denoting columns from table t1. 

And that is exact case we are fighting here.

See testcases added in https://github.com/prestodb/presto/pull/5403/commits/f1403edef98bd3d8b3caef0477a9b853872824eb.

erichwang replied:
I think the main concern here is that the PagesIndex class (which generates the block and position address) is not precisely designed to be used in this way with no columns (@dain can weigh in here with more details) and the leftBlockIndex and leftPosition being used is somewhat bogus. However, it just happens to work in this case.

For the sake of this PR, I think this should be ok for now. But we should work on the longer term fix with the new NestedLoopJoin as soon as possible.

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -447,15 +452,24 @@ private OuterJoinPushDownResult processLimitedOuterJoin(Expression inheritedPred
-            // TODO: consider adding join predicate optimizations to outer joins
+            // See if we can push down any join predicates to the inner side
+            for (Expression conjunct : extractConjuncts(joinPredicate)) {
+                Expression rewritten = innerInference.rewriteExpression(conjunct, not(in(outerSymbols)));
```
Why innerInference here? I think we can use the potentialNullSymbolInference here (as used above) which should give more opportunities for rewrite. Is there an edge case that breaks?

losipiuk replied:
We discussed this code a lot here today. 

Actually it seems you cannot use `potentialNullSymbolInference` here if you are removing the conjunct from `joinConditions` after rewrite. That way you loose information from join clause which is important for proper join semantics.

Extreme example:

```
   LOJ(a=b)
      TS(t1(a))
      TS(t2(b)
```

cannot be rewritten to:

```
   LOJ()
      TS(t1(a))
      FILTER(b=b)
          TS(t2(b))
```

And that is exact behaviour I was getting when I just replaced `innerInference` with `potentialNullSymbolInference`.
Potentially we could use `potentialNullSymbolInference` for extra rewrites which does not remove conjuncts from joinClause and do just pushDown. But I did not want to change that in this patch to be as safe as possible in bugfix PR.

The added commit prevents removal conjuncts from original join condition in case when the original conjunct touched outer symbols. This is just to prevent from getting empty condition in `LOJ` which would imply slow execution currently.
I have second thoughts on that. Probably the better approach would be to leave any condition which can be executed as fast-copiled LookupJoin in join clauses until we have proper implementation of NestedLoopsJoin.
And even then it would probably be not that obvious if removing conjuncts from `joinClause` is always good choice.
As it could for example make data distribution during execution worse.

I will be back by the computer in 2-3 hours. Let me know what you think.

cc: @sogorkis, @pnowojski

erichwang replied:
AH! Yea that will happen if you don't do:

EqualityInference.nonInferrableConjuncts(joinPredicate)
try that instead of:
extractConjuncts(joinPredicate)

you'll see all the other rewrite code paths do that instead

erichwang commented:
Ok just that one last comment in PredicatePushDown

losipiuk commented:
I added yet another fixup commit I was describing in previous message

pnowojski commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -490,6 +490,29 @@ private OuterJoinPushDownResult processLimitedOuterJoin(Expression inheritedPred
+        private boolean shouldRemoveOuterJoinConjunct(Collection<Symbol> outerSymbols, Expression conjunct)
+        {
+
+            if (!(conjunct instanceof ComparisonExpression)) {
```
if conjuct is not `ComparisonExpression` or at least it's not equality comparison and it touches both sides of the table I believe it should still stay? Or not?

pnowojski replied:
ops, it's ok probably, since you are using `innerInference` which doesn't use `joinPredicate`, so by removing it, you don't change semantic.

pnowojski commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -490,6 +490,28 @@ private OuterJoinPushDownResult processLimitedOuterJoin(Expression inheritedPred
 
+        private boolean shouldRemoveOuterJoinConjunct(Collection<Symbol> outerSymbols, Expression conjunct)
+        {
+            if (!(conjunct instanceof ComparisonExpression)) {
```
if conjuct is not ComparisonExpression or at least it's not equality comparison and it touches both sides of the table I believe it should still stay? Or not? Can this change cause a deoptimization? Query like:

```
FILTER (a2 = b2)
  (some complicated query)
    LOJ(a1=b1 AND a2 <= b2)
      TS(t1(a1, a2))
      TS(t2(b1, b2)
```

Could be rewritten by this change to:

```
FILTER (a2 = b2)
  (some complicated operators)
    LOJ(a1=b2)
      TS(t1(a1, a2))
      TS(t2(b1, b2)) and FILTER b2 <= b2
```

which semantically seams ok, but might return more rows from LOJ (which will be filtered out later by FILTER) but still it can cause more work in "(some complicated operators)". Unless we will relay that this `FILTER (a2 = b2)` will be pushed down again in another pass... But if I'm not seriously mistaken (complicated code...) previous version with checking just for `usesOnlySymbols` should be safer.

pnowojski commented:
Can't we write some unit tests for those complicated edge cases for particular re-writters? :/

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -447,15 +452,31 @@ private OuterJoinPushDownResult processLimitedOuterJoin(Expression inheritedPred
+                Expression rewritten = innerInference.rewriteExpression(conjunct, not(in(outerSymbols)));
+                if (rewritten != null) {
+                    innerPushdownConjuncts.add(rewritten);
+                    if (!shouldRemoveOuterJoinConjunct(outerSymbols, conjunct)) {
```
i *think* you might not need this if you try my suggestion above

losipiuk replied:
It would not work (well it would work - but suboptimal). Actually original code looked more or less like what you suggest.
```java
           // See if we can push down any outer or join predicates to the inner side
            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(and(outerEffectivePredicate, joinPredicate))) {
                Expression rewritten = potentialNullSymbolInference.rewriteExpression(conjunct, not(in(outerSymbols)));
                if (rewritten != null) {
                    innerPushdownConjuncts.add(rewritten);
                }
            }
```

But it did not remove anything from `joinPredicate`.
There are two problems with use of `nonInferrableConjuncts`:

1) We want to pushdown to inner side and remove from joinPredicate some conjuncts which are *inferrable* in understanding of `nonInferrableConjuncts`.
E.g `b1 = b2` or `b1 = 5` or `b1 + b2 = 10` (when `b1` and `b2` are columns from inner table).
Those conjuncts would not be subject for rewrite if we used only `nonInferrableConjuncts`.

2) For other *inferrable* conjuncts we want to just do the pushdown but leave conjunct in join clause
E.g `a=b` should be pushed to inner side as `b=5` if we know from inheritedPredicate that `a=5`.
Again this expression would not be taken into consideration if we operate just on `nonInferrableConjuncts`.

The (2) could not be very important - but still beneficial.

But we definitely want (1).
So query like `SELECT .... FROM t1 JOIN t2 on t1.a = t2.b AND t2.c = t2.d` executes quickly without use of `filterFunction` evaluation during execution. This can become less relevant when generated classes for lookup join support filterFunction. But for now I would rather have this.

erichwang replied:
Sorry, so I failed to mention the second part after processing the non inferables. We already do this in other parts of the code (see how we handle inner joins), and I believe we can do something similar here:

```java
        private OuterJoinPushDownResult processLimitedOuterJoin(Expression inheritedPredicate, Expression outerEffectivePredicate, Expression innerEffectivePredicate, Expression joinPredicate, Collection<Symbol> outerSymbols)
        {
            checkArgument(Iterables.all(DependencyExtractor.extractUnique(outerEffectivePredicate), in(outerSymbols)), "outerEffectivePredicate must only contain symbols from outerSymbols");
            checkArgument(Iterables.all(DependencyExtractor.extractUnique(innerEffectivePredicate), not(in(outerSymbols))), "innerEffectivePredicate must not contain symbols from outerSymbols");

            ImmutableList.Builder<Expression> outerPushdownConjuncts = ImmutableList.builder();
            ImmutableList.Builder<Expression> innerPushdownConjuncts = ImmutableList.builder();
            ImmutableList.Builder<Expression> postJoinConjuncts = ImmutableList.builder();
            ImmutableList.Builder<Expression> joinConjuncts = ImmutableList.builder();

            // Strip out non-deterministic conjuncts
            postJoinConjuncts.addAll(filter(extractConjuncts(inheritedPredicate), not(DeterminismEvaluator::isDeterministic)));
            inheritedPredicate = stripNonDeterministicConjuncts(inheritedPredicate);

            outerEffectivePredicate = stripNonDeterministicConjuncts(outerEffectivePredicate);
            innerEffectivePredicate = stripNonDeterministicConjuncts(innerEffectivePredicate);
            joinConjuncts.addAll(filter(extractConjuncts(joinPredicate), not(DeterminismEvaluator::isDeterministic)));
            joinPredicate = stripNonDeterministicConjuncts(joinPredicate);

            // Generate equality inferences
            EqualityInference inheritedInference = createEqualityInference(inheritedPredicate);
            EqualityInference outerInference = createEqualityInference(inheritedPredicate, outerEffectivePredicate);

            EqualityInference.EqualityPartition equalityPartition = inheritedInference.generateEqualitiesPartitionedBy(in(outerSymbols));
            Expression outerOnlyInheritedEqualities = combineConjuncts(equalityPartition.getScopeEqualities());
            EqualityInference potentialNullSymbolInference = createEqualityInference(outerOnlyInheritedEqualities, outerEffectivePredicate, innerEffectivePredicate, joinPredicate);

            // See if we can push inherited predicates down
            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(inheritedPredicate)) {
                Expression outerRewritten = outerInference.rewriteExpression(conjunct, in(outerSymbols));
                if (outerRewritten != null) {
                    outerPushdownConjuncts.add(outerRewritten);

                    // A conjunct can only be pushed down into an inner side if it can be rewritten in terms of the outer side
                    Expression innerRewritten = potentialNullSymbolInference.rewriteExpression(outerRewritten, not(in(outerSymbols)));
                    if (innerRewritten != null) {
                        innerPushdownConjuncts.add(innerRewritten);
                    }
                }
                else {
                    postJoinConjuncts.add(conjunct);
                }
            }
            // Add the equalities from the inferences back in
            outerPushdownConjuncts.addAll(equalityPartition.getScopeEqualities());
            postJoinConjuncts.addAll(equalityPartition.getScopeComplementEqualities());
            postJoinConjuncts.addAll(equalityPartition.getScopeStraddlingEqualities());

            // See if we can push down any outer effective predicates to the inner side
            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(outerEffectivePredicate)) {
                Expression rewritten = potentialNullSymbolInference.rewriteExpression(conjunct, not(in(outerSymbols)));
                if (rewritten != null) {
                    innerPushdownConjuncts.add(rewritten);
                }
            }

            // See if we can push down join predicates to the inner side
            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(joinPredicate)) {
                Expression innerRewritten = potentialNullSymbolInference.rewriteExpression(conjunct, not(in(outerSymbols)));
                if (innerRewritten != null) {
                    innerPushdownConjuncts.add(innerRewritten);
                }
                else {
                    joinConjuncts.add(conjunct);
                }
            }
            // TODO: we can further improve simplifying the equalities by considering other relationships from the outer side
            EqualityInference.EqualityPartition joinEqualityPartition = createEqualityInference(joinPredicate).generateEqualitiesPartitionedBy(not(in(outerSymbols)));
            innerPushdownConjuncts.addAll(joinEqualityPartition.getScopeEqualities());
            joinConjuncts.addAll(joinEqualityPartition.getScopeComplementEqualities())
                    .addAll(joinEqualityPartition.getScopeStraddlingEqualities());

            return new OuterJoinPushDownResult(combineConjuncts(outerPushdownConjuncts.build()),
                    combineConjuncts(innerPushdownConjuncts.build()),
                    combineConjuncts(joinConjuncts.build()),
                    combineConjuncts(postJoinConjuncts.build()));
        }
```

erichwang replied:
So I took a crack at writing the method to see how it would look and this is what I got. After doing so, I think there are 3 main things I want to point out:
-1. We need to be stripping out the non-deterministic join predicate conjuncts from being considered for push down
```java
            joinConjuncts.addAll(filter(extractConjuncts(joinPredicate), not(DeterminismEvaluator::isDeterministic)));
            joinPredicate = stripNonDeterministicConjuncts(joinPredicate);
```
-2. potentialNullSymbolInference is valid and better for rewriting the non-inferrable conjuncts of joinPredicate
-3. We can use the existing equality inference  partitioning behavior to generate which parts of the join predicate equalities to push down:
```java
            EqualityInference.EqualityPartition joinEqualityPartition = createEqualityInference(joinPredicate).generateEqualitiesPartitionedBy(not(in(outerSymbols)));
            innerPushdownConjuncts.addAll(joinEqualityPartition.getScopeEqualities());
            joinConjuncts.addAll(joinEqualityPartition.getScopeComplementEqualities())
                    .addAll(joinEqualityPartition.getScopeStraddlingEqualities());
```
It should be similar to what you were doing before, but without introducing much new code.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java:
```diff
@@ -328,8 +328,15 @@ public PagesHashStrategy createPagesHashStrategy(List<Integer> joinChannels, Opt
-            // temporary hack
+        if (!filterFunction.isPresent() && !joinChannels.isEmpty()) {
+            // todo compiled implementation of lookup join does not support:
+            //  (1) case with join function and the case
```
remove `and the case`

losipiuk commented:
>So I took a crack at writing the method to see how it would look and this is what I got. After doing so, I think there are 3 main things I want to point out:

Thanks for the patch! It is similar but I agree that it is better to use mechanism that are already there.

> -1. We need to be stripping out the non-deterministic join predicate conjuncts from being considered for push down
>
>            joinConjuncts.addAll(filter(extractConjuncts(joinPredicate), not(DeterminismEvaluator::isDeterministic)));
>            joinPredicate = stripNonDeterministicConjuncts(joinPredicate);

Yeah that makes sense. Did not think of this one.

> -2. potentialNullSymbolInference is valid and better for rewriting the non-inferrable conjuncts of joinPredicate

It seems right to me. As long as we are only generating rewrite preconditions based on equality conjuncts - which is the case now.
At least I cannot find any counter example for this approach  :)

> -3. We can use the existing equality inference partitioning behavior to generate which parts of the join predicate equalities to push down:
>
>            EqualityInference.EqualityPartition joinEqualityPartition = createEqualityInference(joinPredicate).generateEqualitiesPartitionedBy(not(in(outerSymbols)));
>            innerPushdownConjuncts.addAll(joinEqualityPartition.getScopeEqualities());
>            joinConjuncts.addAll(joinEqualityPartition.getScopeComplementEqualities())
>                    .addAll(joinEqualityPartition.getScopeStraddlingEqualities());

One issue with this is that we can get rid of all the conjuncts in join condition if they are using only inner side symbols. And get worse code performance as a result due to current execution engine limitations. But my solution also had this issue. I would leave it as it is, as we have to implement better version of `NLJ` and compiled `LookupJoin` pretty soon anyway.

Concerning your `todo`:
```java
 // TODO: we can further improve simplifying the equalities by considering other relationships from the outer side
```
This make sense. I will add separate short PR for that one but lets leave it outside of this PR so we get rid of release blocker  

And just one general question.
All this code uses `not(in(outerPredicate))` as equivalend of `in(innerPredicate)`. Is that actually an equivalence? I guess there is but could you point me to the code part which makes sure no extra pradicates which should not be in scope are not present.

I left your patch as separate commit. Feel free to either squash it or leave as is. The tests seem to pass.

erichwang commented:
> One issue with this is that we can get rid of all the conjuncts in join condition if they are using only inner side symbols. And get worse code performance as a result due to current execution engine limitations. But my solution also had this issue. I would leave it as it is, as we have to implement better version of NLJ and compiled LookupJoin pretty soon anyway.

Yea, I think it's better now to have the right plan even though its suboptimal for the short term solution. As you mentioned, it should be always better for the proper NLJ approach.

> Concerning your todo:
This make sense. I will add separate short PR for that one but lets leave it outside of this PR so we get rid of release blocker

I would recommend holding off on this for now. My feeling is that to do it properly, we will need to make some architectural changes to the EqualityInference to do it cleanly -- which might get nasty.

> And just one general question.
All this code uses not(in(outerPredicate)) as equivalend of in(innerPredicate). Is that actually an equivalence? I guess there is but could you point me to the code part which makes sure no extra pradicates which should not be in scope are not present.

Yes, not-in-outer should be equal to in-inner. We know this for a fact by the definition of how joins operate. For the context of this method, you can assume that to be true.

erichwang commented:
Merged, thanks!
